docs/hugo/content/design/ADR-2022-11-Change-Detection.md

```diff
@@ -4,79 +4,115 @@ title: '2022-11: Change Detection'
+This is something that we can get _good enough_, but it will take iteration to get actually correct. Practically speaking, it's hard to know what we're doing well or poorly with diffs because, for example, with the normalization case, if you look at something like MySQLServer, it has maybe 30 user-settable fields. If one of those fields normalizes certain values when set, how would we ever notice?
 
-For properties we use in our recorded tests, we'll see issues when we go to re-record them, but for properties we don't use in our tests, we'll never know. 
+For properties we use in our recorded tests, we'll see issues when we go to re-record them, but for properties we don't use in our tests, we'll never know.
 
 This points to a need to log which fields are detected as changed, so that we can review the ASO logs and identify when we're getting the diff comparison wrong.
 
+### Issue: Storage of known-safe differences
+
+If we are going to track the differences between resource-as-PUT and resource-as-created-in-Azure, we need to store those differences somewhere. For most resources, this will be a very small delta, but we need to allow for the possibility that it could be large.
+
+**Possible mitigations**:
+
+* Store as an annotation on the resource. This is simple but has a distinct size limit (256KiB for _all annotations combined_). This is probably sufficient for most resources, but we need to be aware of the limit and decide what to do if we exceed it.
+
+* Capture a hash of the diff between resource-as-PUT and resource-as-created-in-Azure and store that as an annotation. This will have a fixed size but means we won't have the actual diff itself. Any change at all will result in a different hash, so we'll likely be performing a PUT to reconcile more often, and we won't be able to log the actual differences for troubleshooting purposes.
+
+* Add `operatorStatus` to the status of the resource and use that to capture a map of the changes found between resource-as-PUT and resource-as-created-in-Azure. This avoids the size limit of annotations and makes differences easy to inspect and log.
+
+### Issue: Performance
+
+What's the CPU overhead of doing a GET and diff on every reconciliation compared to just doing a PUT?
+
+We'll be adding another network round trip (with the associated bandwidth and latency costs) for every reconciliation, plus the CPU cost of doing the diff itself. We need to ensure that this doesn't cause us to exceed our resource limits in the controller pod, especially when reconciling large numbers of resources.
+
 ## Decision
 
-TBC
+Recommendation: Implement differencing as follows:
+
+* Update our code generator to add a `DiffWith()` method to every spec type that accumulates a list of the differences found
```

Comment:
Can we expand on the actual algorithm here some more? We cover above for example the fact that we'll cover the status type back to spec and discard the readonly fields, but that's not really covered here.

Reply:
Expanded with considerably more detail.